### PR TITLE
[P1] Build read-only repo and GitHub context broker for agentic review

### LINKED ISSUE #57: [P1] Build read-only repo and GitHub context broker for agentic review
Part of #331.

## Problem
Cerberus still relies on a prompt-heavy review surface: we bootstrap a diff file, inject large chunks of context into the prompt, and ask reviewers to reason from that static bundle. Even after the large-PR slice work, this keeps the runtime brittle in three ways:
- reviewers cannot fetch the exact adjacent context they need when a diff hint is ambiguous
- PR text, linked issue text, and review comments are mixed into the same prompt surface instead of being treated as untrusted inputs retrieved on demand
- the system is hard to evaluate because we do not have a first-class contract for what context the reviewer may read and how it should retrieve it

The repo already has partial ingredients for a better shape: repo read/search tools, `github-read`, and runtime helpers in `scripts/lib/github_platform.py`. What is missing is a deliberate read-only context broker for agentic review.

## Outcome
Cerberus reviewers get a typed, read-only context surface for agentic exploration of the diff, repo, and GitHub discussion.

That surface should let reviewers:
- list changed files and changed-file metadata
- read file content at the relevant ref
- fetch the diff for one file or a bounded slice of files
- retrieve linked issue and PR discussion context through structured reads
- search the local repo without falling back to unrestricted shell or raw `gh` output

## Context
This is the foundation for the broader agentic-review direction discussed in March 2026:
- give reviewers more autonomy to explore the codebase and PR context
- keep tool surfaces typed, read-only, and auditable
- avoid relying on giant procedural prompts as the main source of review quality

Related work:
- #333 security/dataflow recall hardening
- #334 large-PR timeout blind-spot reduction
- #335 lifecycle/state-machine challenger lane
- #344 configurable reviewer benches with router-selected execution
- #346 routed single-matrix execution

## Acceptance Criteria
- [ ] [behavioral] Given a reviewer runtime on this branch, when a reviewer needs adjacent code or GitHub context, then it can retrieve that context through explicit read-only tools instead of only through pre-injected prompt text.
- [ ] [behavioral] Given a PR with linked issue discussion and review comments, when the reviewer asks for that context, then the returned data is structured, bounded, and marked as untrusted evidence rather than silently mixed into the base prompt.
- [ ] [test] Given the reviewer profile and tool broker configuration, when repo tests inspect them, then they verify a typed read-only contract for diff, file, search, and GitHub context retrieval.
- [ ] [test] Given future changes try to route reviewers back through unrestricted shell or raw `gh` access for context gathering, when contract tests run, then they fail.
- [ ] [command] Given the implementation on this branch, when `python3 -m pytest tests/test_reviewer_profiles.py tests/test_github_platform.py tests/test_review_prompt_project_context.py -q` runs, then the tool-contract regressions pass.

## Touchpoints
- `defaults/reviewer-profiles.yml`
- `pi/extensions/github-read.ts`
- `scripts/lib/github_platform.py`
- `templates/review-prompt.md`
- `tests/test_reviewer_profiles.py`
- `tests/test_github_platform.py`

## Verification
```bash
python3 -m pytest tests/test_reviewer_profiles.py tests/test_github_platform.py tests/test_review_prompt_project_context.py -q
make test
```

## Boundaries
- Do not add write-capable GitHub tools in this lane.
- Do not expose a generic unrestricted `gh` shell surface to reviewers.
- Do not redesign bench selection or workflow orchestration here; this issue is about context access, not scheduling.
- Do not collapse all context into one giant prompt as a fallback for missing tool work.

## Related
- This issue upgrades the old MCP prototype idea into an active OSS reviewer-runtime foundation.
- Prompt simplification and eval hardening should land as follow-on issues, not as hidden scope inside this one.

### Problem
Cerberus reviewers still depend on a static prompt bundle plus generic filesystem tools to understand a PR. That makes context gathering inconsistent, mixes trusted bootstrap and untrusted GitHub discussion too loosely, and leaves no explicit contract for how reviewers should retrieve adjacent code, diff slices, and issue discussion.

### Users
**Primary**: Cerberus maintainers tuning reviewer quality and reliability.
**Secondary**: Downstream repos relying on Cerberus reviewers to fetch the right context without brittle prompt stuffing.

### Recommended Approach
Add a typed, read-only repo context tool alongside the existing typed GitHub read tool, then route reviewer guidance and profile configuration through those bounded tools. Keep prompt bootstrap small, but let reviewers fetch the exact file, diff, search, and GitHub context they need through auditable contracts.

### Intent Contract
- Intent: Reviewers retrieve diff, file, search, and GitHub context through explicit read-only tools instead of depending only on pre-injected prompt text.
- Success Conditions: Reviewer runtime exposes a typed repo broker and typed GitHub broker; prompts point reviewers at those tools; regression tests fail if the contract regresses.
- Hard Boundaries: No write-capable GitHub tools, no unrestricted shell/raw `gh` surface for reviewer context gathering, no workflow-orchestration redesign.
- Non-Goals: Bench routing changes, prompt rewrites unrelated to context access, or broad runtime refactors outside the context broker boundary.

### User Stories
- As a reviewer agent, I want to ask for one file slice or one diff slice so that I do not reason from an oversized prompt bundle.
- As a maintainer, I want tests to lock the allowed context surface so future changes cannot quietly reintroduce unsafe access patterns.

### Success Metrics
| Metric | Target | How Measured |
| --- | --- | --- |
| Repo context retrieval contract | Present | Contract tests for repo broker actions and reviewer profile wiring |
| GitHub context trust boundary | Preserved | Prompt/tests keep GitHub discussion on explicit `github_read` path |
| Context regressions | Blocked | Tests fail if broker config drops typed repo/github context or reintroduces raw access |

### Non-Goals
- Add GitHub write tools to reviewer runtime.
- Replace the whole reviewer runtime with a new execution engine.
- Fold all repo and GitHub context back into the prompt as a fallback-first design.

### Boundaries
- Do NOT modify reviewer verdict schema in this lane.
- Do NOT add generic shell access for reviewers.
- Do NOT redesign project routing, waves, or bench selection.

### PR Intent Reference
- Implementation PR must link this issue and describe how the typed repo/github context contract is enforced.

### Approach
Introduce a new `repo_read` Pi extension that reads from the existing `review-run` contract and exposes bounded actions for changed-file metadata, file reads, diff reads, and repo search. Keep GitHub discussion retrieval in the existing `github_read` extension, update reviewer profiles so both extensions are part of the base contract, and tighten tests around prompt guidance and profile wiring.

### Files to Modify/Create
- `defaults/reviewer-profiles.yml` — add the repo context broker extension to the shared reviewer runtime contract.
- `pi/extensions/repo-read.ts` — implement typed read-only repo context actions backed by the review-run contract and workspace filesystem.
- `templates/review-prompt.md` — instruct reviewers to use explicit repo/github tools for context retrieval.
- `tests/test_reviewer_profiles.py` — lock the shared extension contract.
- `tests/test_review_prompt_project_context.py` — verify prompt guidance references repo/github tools.
- `tests/test_repo_read_contract.py` — assert declared repo broker actions and guardrails.
- `tests/extensions/repo-read.test.ts` — exercise extension behavior and bounded reads/search.

### Interfaces
- `repo_read(action="list_changed_files")` — returns changed-file metadata parsed from the review diff.
- `repo_read(action="read_file", path, startLine?, endLine?)` — returns bounded file slices from the checked-out workspace.
- `repo_read(action="read_diff", path?, limit?)` — returns the full diff or selected file diffs from the review-run diff artifact.
- `repo_read(action="search_repo", query, pathPrefix?, limit?)` — returns bounded repo search hits rooted at the review workspace.

### Implementation Sequence
1. Add the repo broker extension and wire it into the base reviewer profile.
2. Update reviewer prompt guidance to prefer `repo_read` plus `github_read` instead of prompt-only context.
3. Add contract and integration tests for the new extension and its profile/prompt wiring.
4. Run targeted tests, then broader validation if the lane stays green.

### Testing Strategy
- Python contract tests for reviewer profiles and prompt rendering.
- Python text-contract tests for the repo-read extension surface.
- Node extension tests for repo-read behavior, path bounds, diff parsing, and search behavior.

### Risks & Mitigations
- Risk: Repo reads escape the checked-out workspace.
  Mitigation: Resolve all paths against `workspace_root` from `review-run` and reject escapes.
- Risk: Search or diff reads become unbounded.
  Mitigation: Enforce line/result/file limits in the extension contract.
- Risk: Prompt guidance drifts away from the contract.
  Mitigation: Add explicit prompt regression tests.

### PR DESCRIPTION
**Pull request by moomooskycow**
## Reviewer Evidence
- Start here: [issue-57 reviewer evidence](https://github.com/misty-step/cerberus/blob/cx/issue-57-repo-read-broker/walkthrough/issue-57-reviewer-evidence.md)
- Direct artifact: [pr-57 walkthrough](https://github.com/misty-step/cerberus/blob/cx/issue-57-repo-read-broker/artifacts/pr-57-walkthrough.md)
- Walkthrough notes: internal-only change; evidence is command output and contract coverage rather than UI/video capture.
- Fast claim: reviewer runtimes now have a typed `repo_read` broker for local context, hardened after review to block symlink escapes, reject invalid file ranges, bound repo-search reads, and preserve odd diff paths.

## Why This Matters
- Problem: reviewers could fetch live GitHub discussion via `github_read`, but local repo context still depended on generic file tools and the pre-rendered diff prompt.
- Value: this makes local context retrieval explicit, bounded, auditable, and now pressure-tested against the boundary/parsing edge cases the external reviewers found.
- Why now: issue #57 is the foundation lane for moving review quality away from prompt-heavy context stuffing.
- Issue: Closes #57

## Trade-offs / Risks
- Value gained: typed changed-file metadata, bounded file/diff reads, and repo search in one shared contract, plus a reusable reviewer-delta triage skill for comparing Cerberus against other review bots.
- Cost / risk incurred: one more runtime extension, a small triage skill surface, and a simple text-search implementation that is intentionally conservative rather than optimized.
- Why this is still the right trade: the repo needed a stable contract and explicit hardening before pursuing richer retrieval.
- Reviewer watch-outs: repo search is still straightforward recursive text scanning; this lane intentionally stops short of symbol-aware or AST-aware navigation.

## What Changed
This PR adds a new `repo_read` extension, loads it in the shared reviewer runtime profile, updates prompt guidance to split local repo access from GitHub discussion access, adds a reviewer-delta triage skill for bot-vs-Cerberus comparison, and hardens the broker after review by fixing symlink, invalid-range, search-bound, and odd diff-path edge cases with regression coverage.

### Base Branch
```mermaid
graph TD
  Prompt["Prompt + diff file"] --> GenericTools["Generic read/grep/find/ls tools"]
  Prompt --> GitHubRead["github_read"]
  GenericTools --> LocalContext["Unstructured local repo exploration"]
  GitHubRead --> GitHubContext["Typed GitHub discussion reads"]
```

### This PR
```mermaid
graph TD
  Prompt["Prompt + diff file"] --> RepoRead["repo_read"]
  Prompt --> GitHubRead["github_read"]
  RepoRead --> LocalContext["Typed changed-files / file slices / diff slices / repo search"]
  GitHubRead --> GitHubContext["Typed GitHub discussion reads"]
```

### Architecture / State Change
```mermaid
graph TD
  ReviewRun["review-run contract"] --> RepoRead["repo_read extension"]
  RepoRead --> Diff["diff_file"]
  RepoRead --> Workspace["workspace_root"]
  Profiles["reviewer-profiles.yml"] --> RepoRead
  Prompt["review-prompt.md"] --> RepoRead
  Tests["contract + extension tests"] --> RepoRead
  Delta["reviewer-delta-triage skill"] --> Cerberus["Cerberus hardening loop"]
```

Why this is better:
- Local repo context now has the same explicit-tool shape GitHub context already had.
- The runtime contract is narrower and more testable than relying on generic filesystem exploration.
- The new edge-case tests make the “bounded and auditable” claim real instead of aspirational.

<details>
<summary>Intent Reference</summary>

## Intent Reference
Issue #57 asks for a typed, read-only context surface so reviewers can retrieve changed files, adjacent code, diff slices, and GitHub discussion through explicit tools instead of relying only on pre-injected prompt text.

Source issue: [#57](https://github.com/misty-step/cerberus/issues/57)

</details>

<details>
<summary>Changes</summary>

## Changes
- Added `pi/extensions/repo-read.ts` with `list_changed_files`, `read_file`, `read_diff`, and `search_repo`.
- Hardened `repo_read` against symlink escapes, out-of-bounds file slices, conservative truncation reporting, bounded search reads, and odd diff paths containing ` b/`.
- Loaded `repo_read` from `defaults/reviewer-profiles.yml` beside `github_read`.
- Updated `templates/review-prompt.md` to direct reviewers to `repo_read` for local context.
- Added regression coverage in `tests/test_repo_read_contract.py`, `tests/extensions/repo-read.test.ts`, `tests/test_reviewer_profiles.py`, and `tests/test_review_prompt_project_context.py`.
- Added reviewer evidence in `artifacts/pr-57-walkthrough.md` and `walkthrough/issue-57-reviewer-evidence.md`.
- Added `.agents/skills/reviewer-delta-triage/*` plus `artifacts/pr-383-reviewer-delta-triage.md` to codify bot-vs-Cerberus comparison.
- Updated `AGENTS.md` and `CLAUDE.md` to treat `context-engineering`, `llm-infrastructure`, and `harness-engineering` as the default Cerberus operating frame.

</details>

<details>
<summary>Alternatives Considered</summary>

### Option A — Do nothing
- Upside: no new extension surface.
- Downside: reviewers keep mixing prompt stuffing with generic local exploration.
- Why rejected: it does not satisfy the issue’s core contract goal.

### Option B — Push more local context into the prompt
- Upside: no runtime-tool work.
- Downside: larger prompts, blurrier trust boundaries, weaker testability.
- Why rejected: it doubles down on the shape this issue is trying to replace.

### Option C — Add a typed repo broker
- Upside: explicit, bounded, testable local context access.
- Downside: adds one runtime extension to maintain.
- Why chosen: it matches the intended agentic-review architecture with the smallest reversible diff.

</details>

<details>
<summary>Acceptance Criteria</summary>

## Acceptance Criteria
- [x] [behavioral] Reviewer runtime can retrieve adjacent local repo context through explicit read-only tools instead of only through pre-injected prompt text.
- [x] [behavioral] GitHub issue and PR discussion context remains structured, bounded, and explicitly tool-retrieved.
- [x] [test] Reviewer profile and broker configuration verify a typed read-only contract for diff, file, search, and GitHub context retrieval.
- [x] [test] Contract tests fail if future changes try to route reviewers back through unrestricted shell or raw `gh` access for context gathering.
- [x] [command] `python3 -m pytest tests/test_reviewer_profiles.py tests/test_github_platform.py tests/test_review_prompt_project_context.py -q`

</details>

<details>
<summary>Manual QA</summary>

## Manual QA
1. Run `npx -y tsx --test tests/extensions/github-read.test.ts tests/extensions/repo-read.test.ts`.
   Expected: `18` tests pass.
2. Run `python3 -m pytest tests/test_repo_read_contract.py tests/test_reviewer_profiles.py tests/test_review_prompt_project_context.py -q`.
   Expected: all targeted contract and prompt lanes pass.
3. Run `python3 -m pytest tests/test_reviewer_profiles.py tests/test_github_platform.py tests/test_review_prompt_project_context.py -q`.
   Expected: all targeted runtime-contract tests pass.
4. Run `make validate`.
   Expected: full repo test, lint, and shellcheck gates pass.

</details>

<details>
<summary>Walkthrough</summary>

## Walkthrough
- Renderer: markdown evidence for an internal runtime change.
- Artifact: [artifacts/pr-57-walkthrough.md](https://github.com/misty-step/cerberus/blob/cx/issue-57-repo-read-broker/artifacts/pr-57-walkthrough.md)
- Claim: the branch adds a typed repo broker, hardens the review-surfaced edge cases, including rename headers, and proves it with extension tests plus full validation.
- Before / After scope: reviewer runtime context access only; no workflow-routing redesign.
- Persistent verification: `make validate`
- Residual gap: repo search is intentionally simple text scanning for now.

</details>

<details>
<summary>Before / After</summary>

## Before / After
Before: reviewer runtimes had a typed GitHub context tool but no typed local repo context broker.
After: reviewer runtimes have both `repo_read` and `github_read`, with prompt guidance, edge-case hardening, and tests locking the split.

Screenshots are not needed because this PR changes runtime contracts and tests, not a user-facing UI.

</details>

<details>
<summary>Test Coverage</summary>

## Test Coverage
- `tests/extensions/repo-read.test.ts`
- `tests/test_repo_read_contract.py`
- `tests/test_reviewer_profiles.py`
- `tests/test_review_prompt_project_context.py`
- `tests/test_github_platform.py`

Gap: no symbol-aware search behavior is introduced or tested in this lane because the broker intentionally ships a bounded text-search primitive only.

</details>

## Summary by CodeRabbit

* **New Features**
  * Introduced local repository context access for reviewers to list changed files, read bounded file slices, inspect diffs, and search the checked-out repo.
  * Added a CLI tool to collect PR review surface data and a Reviewer Delta Triage skill to analyze PR review deltas.

* **Documentation**
  * New walkthroughs, reviewer evidence, and skill docs updating review workflows and guidance.

* **Tests**
  * Added comprehensive JS/Python tests and contract checks validating repo-access behavior and boundary protections.

### CODE REVIEW

File: pi/extensions/repo-read.ts [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+
+function resolveWorkspacePath(workspaceRoot: string, relativePath: string): string {
+	const resolvedRoot = path.resolve(workspaceRoot);
+	const candidate = path.resolve(resolvedRoot, relativePath);
+	const relative = path.relative(resolvedRoot, candidate);
+	if (relative === ".." || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
+		throw new Error("path escapes workspace root");
+	}
+	return candidate;
+}
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply pi/extensions/repo-read.ts]: Addressed in commit ffa4826. `resolveWorkspacePath` now does the lexical containment check and then re-checks the canonical real path against the canonical workspace root with `fs.realpathSync`, so symlinked files or directories can no longer escape `workspace_root`. I added regression tests for both `read_file` and `search_repo` via a symlinked directory. Verified with `npx -y tsx --test tests/extensions/github-read.test.ts tests/extensions/repo-read.test.ts`, `python3 -m pytest tests/test_repo_read_contract.py tests/test_reviewer_profiles.py tests/test_review_prompt_project_context.py -q`, and `make validate`.

File: pi/extensions/repo-read.ts [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+function resolveWorkspacePath(workspaceRoot: string, relativePath: string): string {
+	const resolvedRoot = path.resolve(workspaceRoot);
+	const candidate = path.resolve(resolvedRoot, relativePath);
+	const relative = path.relative(resolvedRoot, candidate);
+	if (relative === ".." || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
+		throw new Error("path escapes workspace root");
+	}
+	return candidate;
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply pi/extensions/repo-read.ts]: Addressed in commit ffa4826. `resolveWorkspacePath` now does the lexical containment check and then re-checks the canonical real path against the canonical workspace root with `fs.realpathSync`, so symlinked files or directories can no longer escape `workspace_root`. I added regression tests for both `read_file` and `search_repo` via a symlinked directory. Verified with `npx -y tsx --test tests/extensions/github-read.test.ts tests/extensions/repo-read.test.ts`, `python3 -m pytest tests/test_repo_read_contract.py tests/test_reviewer_profiles.py tests/test_review_prompt_project_context.py -q`, and `make validate`.

File: pi/extensions/repo-read.ts [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+function buildFileSlice(
+	filePath: string,
+	relativePath: string,
+	startLineRaw: number | undefined,
+	endLineRaw: number | undefined,
+) {
+	const text = readTextFile(filePath);
+	const lines = text.split("\n");
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply pi/extensions/repo-read.ts]: Addressed in commit ffa4826. `buildFileSlice` now rejects `startLine` values beyond the file length before computing `endLine`, so the tool no longer returns contradictory `startLine > endLine` payloads. I added an explicit EOF regression test in `tests/extensions/repo-read.test.ts`. Verified with `npx -y tsx --test tests/extensions/github-read.test.ts tests/extensions/repo-read.test.ts`, `python3 -m pytest tests/test_repo_read_contract.py tests/test_reviewer_profiles.py tests/test_review_prompt_project_context.py -q`, and `make validate`.
